### ISSUE TITLE
Exposure Logging is in false "restricted" state without internet connection

### ISSUE THREAD

**Issue by kbobrowski** [corona-warn-app/cwa-backlog#1]
:warning: **edit: originally posted issue below may be misleading users to believe that they may safely keep internet off from time to time, and that exposure notification will be updated when internet connectivity is recovered. This should be the case, but might not be currently, see [this comment](https://github.com/corona-warn-app/cwa-documentation/issues/262#issuecomment-646490809). I will keep this header until the issue described in this comment is clarified.** :warning:

Note: this is a different issue to corona-warn-app/cwa-app-android#491 and can be solved independently.

## Describe the bug

When there is no internet access exposure logging will go to "restricted" state, the same state which appears when Bluetooth is off. But even without internet access phones will still receive / send Bluetooth beacons, and they can be matched later on with Diagnosis Keys (once internet access is enabled again).

This looks like sub-optimal UI/UX design, and leads to following issues:

- a person without mobile data enabled might believe that it does not make sense to take the phone or turn on Bluetooth when going e.g. grocery shopping, since exposure logging is not active anyway without internet access
- a person taking a flight might be led to believe that it does not make sense to turn on Bluetooth (after both internet access and Bluetooth have been turned off by airplane mode), since without internet the app won't work anyway (as a side note - it's possible to safely switch on Bluetooth after airplane mode was activated - some people may not know about this, perhaps it'd be good to indicate it in the app as well)
- **user has no way to opt-out (stop transmitting BLE beacons) once phone has internet connection disabled**

## Expected behaviour

It should be clear for the user that even without internet access app is working (it just needs internet access from time to time), and it's still worth to always carry a phone / keep Bluetooth on even in a situations without internet access.

**Additionally, user should be able to opt-out at any time, regardless of whether internet access is enabled, as stated in privacy policy:**

> To withdraw your consent to the exposure logging feature, you can disable the feature
using the toggle switch in the App or delete the App.

## Possible Fix

Better UI/UX design with two separate states:
- state indicating if exposure logging is on (this requires Bluetooth)
- state indicating if exposure risk score is up-to-date (this requires internet)

**Comment by aurisnoctis** [comment]
@kbobrowski 
> This issue may be related to issues discussed in [corona-warn-app/cwa-documentation#236](https://github.com/corona-warn-app/cwa-documentation/issues/236) (see [this comment](https://github.com/corona-warn-app/cwa-documentation/issues/236#issuecomment-646238207)) - essentially **if internet connectivity is down in the short one-minute time window it may be that exposure status is not updated until next day**, and then again it has to be active in a short time window during next day.

If internet was not available at some point when the app tried to update the risk status and the app shows accordingly that risk status was updated e.g. `day before yesterday` --> Can I expect that if I have internet connection continuously from now on that this 'skipped' day will not be entirely lost for matching? For the example if the last risk update is from 'day before yesterday' and `today` the risk update works correctly as indicated by the updated status in the app, will it include matching of encounters for **yesterday**?

**Comment by daimpi** [comment]
@tkowark Given that there will be a [fix](https://github.com/corona-warn-app/cwa-app-ios/pull/970) for this issue under iOS in CWA 1.3, will we get something similar for Android?

I can reliably reproduce the issue:
1. Open CWA
2. Keep mobile data on
2. turn off Wifi -> everything seems fine, CWA is fine with mobile data
3. turn Wifi on again 
4. After the phone is connected to Wifi again, the "no internet connection" status appears in the app.
5. Switching out of CWA and back in again, fixes the "no internet connection" status

My device: Samsung Galaxy S8 (Android 9) with CWA 1.2.1

**Comment by harambasicluka** [comment]
The Internet status will be removed with https://github.com/corona-warn-app/cwa-app-android/pull/1096

Thanks for your patients.

Cheers,
Luka

**Comment by Ein-Tim** [comment]
Should this problem be solved with the version 1.3.0?

**Comment by Ein-Tim** [comment]
Could please somebody from the devs come back to us and give an update about this Issue.
Will [this](https://github.com/corona-warn-app/cwa-backlog/issues/1#issuecomment-687175144) contradiction be fixed?

**Comment by thomasaugsten** [comment]
You should receive an error popup with 1.3.1

**Comment by Ein-Tim** [comment]
Yeah, I do but this is even more confusing. The user will get
a) confused
b) think that the App isn't working without Internet connection...

Imho this is suboptimal and has to be improved.

**Comment by thomasaugsten** [comment]
This is the fact the initial start of the app is not working without internet

**Comment by Ein-Tim** [comment]
Yeah but the user will think the whole App is not working...
And please see my screenshot from [above](https://github.com/corona-warn-app/cwa-backlog/issues/1#issuecomment-687175144), don't SAP and the RKI see a contradiction here?

**Comment by daimpi** [comment]
> I can reliably reproduce the issue:
> 
>  1. Open CWA
> 
>  2. Keep mobile data on
> 
>  3. turn off Wifi -> everything seems fine, CWA is fine with mobile data
> 
>  4. turn Wifi on again
> 
>  5. After the phone is connected to Wifi again, the "no internet connection" status appears in the app.
> 
>  6. Switching out of CWA and back in again, fixes the "no internet connection" status
> 
> 
> My device: Samsung Galaxy S8 (Android 9) with CWA 1.2.1

Same as @MikeMcC399 (https://github.com/corona-warn-app/cwa-app-android/issues/1141#issuecomment-712193621) I can't reproduce this issue anymore with CWA 1.5 under Android 🎉.

**Comment by Ein-Tim** [comment]
Thats nice @daimpi, could you share a Screenshot what the Android App is now showing without Internet?

**Comment by daimpi** [comment]
Yes sure, this is how my phone looks on airplane mode with Bluetooth enabled:

<img src="https://user-images.githubusercontent.com/22129968/96469354-69a23b00-122d-11eb-88bb-4a09778ed30e.png" width="300">

**Comment by Ein-Tim** [comment]
Thats how I think the App should behave without Internet. On iOS, the behaviour is still [the same](https://github.com/corona-warn-app/cwa-backlog/issues/1#issuecomment-687175144) as in the Versions before... And seems like they don't want to change anything here...

**Comment by daimpi** [comment]
> On iOS, the behaviour is still the same as in the Versions before... And seems like they don't want to change anything here...

That's a pity 🙁. 

@thomasaugsten is there anything planned wrt adopting this improvement also for iOS?

**Comment by Ein-Tim** [comment]
Okay interesting...
Now my App looks like @daimpi's one without Internet... Seems like they fixed it for iOS too.
So if anybody else could confirm this (cc @ndgendogo) I think this Issue can be closed 👍

**Comment by daimpi** [comment]
That's great to hear 😊 🎉 

> So if anybody else could confirm that the behaviour on iOS is also correct now, I think this Issue can be closed 👍

@ndegendogo could you maybe check whether you can confirm @Ein-Tim's findings on your device? 🙂

**Comment by ndegendogo** [comment]
@daimpi on the first glance it looks fine. However, I would like to try a few more tests in Wifi later.
(Recently, I had a very strange and annoying behaviour with cwa 1.3.2 when I had Wifi without internet - not sure if I can reproduce this condition easily).

**Comment by daimpi** [comment]
@ndegendogo sounds good 👍 

Looking forward to your results 🙂.

**Comment by ndegendogo** [comment]
So, not 100% sure what is the expected / desired behaviour.
With Wifi and mobile data enabled, but no internet coverage I get an error popup message. Same if I have mobile data disabled and Wifi enabled but no coverage.
I can acknowledge the error, and then see the main screen and my risk status. The message pops up again each time when I open the app or return from a submenu to the main screen.
The main screen shows clearly that "Risiko Ermittlung aktiv" (this was the original focus of this ticket, yes?)
When I disable Wifi: no error popup.

**Comment by daimpi** [comment]
@ndegendogo that doesn't look like it's intended. Do I understand this correctly: you get this message when you don't have internet coverage but have the respective networks (wifi and/or mobile data) enabled (but they are unable to provide internet for some reason)? 

So could this be reproduced in the following way?
1. remove your home wifi password from your device
2. disable mobile data, while leaving wifi enabled (is this necessary? I thought this would be needed b/c simulating no mobile coverage might be hard)
3. error appears

Or is there maybe an easier way to reproduce this?

**Comment by ndegendogo** [comment]
easiest way to reproduce: 
- disable mobile data
- keep Wifi enabled
- go for a walk (out of Wifi coverage)

It may be that iOS behaves differently than Android.

**Comment by daimpi** [comment]
@ndegendogo 
> easiest way to reproduce:
> 
> * disable mobile data
> 
> * keep Wifi enabled
> 
> * go for a walk (out of Wifi coverage)
> 
> 
> It may be that iOS behaves differently than Android.

I couldn't reproduce this under Android, I didn't get any error message and "exposure logging active" stayed on all the time.

@Ein-Tim can you reproduce the error on your iOS device?

**Comment by Ein-Tim** [comment]
@daimpi 
Can reproduce this Issue from @ndegendogo, but on my phone it looks like this:

<img src="(https://user-images.githubusercontent.com/67682506/96677155-4bdcef00-136f-11eb-9c3c-830c938ebbcd.png)" width="300">

**Comment by ndegendogo** [comment]
@Ein-Tim but the wording is *still* very confusing and contradictory, IMO.

Maybe it is worth a separate ticket, but: what exactly do they mean with "Risiko-Ermittlung"? The BLE part? The daily key download and matching? There should be different and clearly distinguishing words for these two procedures. Especially because they have different technical preconditions, and the user may want to deactivate one but not the other.

**Comment by Ein-Tim** [comment]
@ndegendogo 
Yes, the wording which you can see on the last screenshot is very confusing. But I mean it got a bit better because sometimes the App also shows [the correct screen](https://github.com/corona-warn-app/cwa-backlog/issues/1#issuecomment-712594570). 

The little Box on the Top refers to the Exchange of BLE Keys and the big Box in the Middle refers to the downloading of new Keys from the Server. (As of my understanding) 
I agree with you @ndegendogo, these two processes should be separated better...

**Comment by ndegendogo** [comment]
My guess on the error popup: maybe a (failed) attempt to download the keys for the risk check.
And I appreciate **a lot** that I can trigger this procedure manually;  with a broken background task for several days you start to panic ...
Only there is no need to try this again and again on every return from a submenu. Slightly annoying; but not the biggest issue here.

**Comment by Ein-Tim** [comment]
@ndegendogo @daimpi 
Sooo, during the day I saw:
Nothing really changed imho, the App still shows [this](https://github.com/corona-warn-app/cwa-backlog/issues/1#issuecomment-713313608) contradictory screen if the phones isn't connected to the Internet.
So this still has to be improved, even when it looks like that in some situations the App now shows the right screen...
(Tbh, I could not (yet) find any pattern when the App works and when not...)

**Comment by daimpi** [comment]
So overall the issue seems to be fine on Android now, but is definitely not resolved on iOS.

**Comment by ndegendogo** [comment]
> any pattern when the App works and when not...

My guess on the pattern is:
- when Wifi or mobile data is deactivated by user, iOS (and cwa) knows in advance that there will be no internet on this path. No attempt to connect, no error message.
- when Wifi or mobile data is enabled by user, cwa tries to connect to the internet. 
- if this attempt fails -> error message.

@Ein-Tim would this be consistent with what you see?

**Comment by Ein-Tim** [comment]
@ndegendogo 
Not really, no, I was sitting in class today with my phone in Flight Mode (but Bluetooth enabled) and CWA still showed the "wrong" contradictory screen... 

Again now, I deactivated Mobile Data 

<img src="https://user-images.githubusercontent.com/67682506/96738044-9a65aa00-13be-11eb-852a-cc4561fe340d.png" width="200">

and opened the App, this screen was shown: 

<img src="https://user-images.githubusercontent.com/67682506/96738140-b49f8800-13be-11eb-868c-af02c08aea49.png" width="200">

**Comment by ndegendogo** [comment]
@Ein-Tim 
So maybe different behavior of the different iOS versions? (I am on 13.7)

**Comment by Ein-Tim** [comment]
I'm on 14.1

**Comment by ndegendogo** [comment]
Yesterday I also updated to 14.1. Mainly because of #1106.
So, I'll repeat my tests and will report the results.

**Comment by ndegendogo** [comment]
With: cwa 1.5.0 (8) / iOS 14.1 / iPhone 8 
I could still reproduce the error message, like yesterday with iOS 13.7. (maybe it was slightly harder).
I did not see [that message](https://github.com/corona-warn-app/cwa-backlog/issues/1#issuecomment-713639685)

**Comment by Ein-Tim** [comment]
Mhh @ndegendogo 
I think it definetly got better but I can still sometimes repoduce [this screen](https://github.com/corona-warn-app/cwa-backlog/issues/1#issuecomment-713639685).

@Oliver-Zimmerman @hermesmar (tagging you because you are asigned to this ticket)
Are there any improvements planed for further releases regarding the behaviour of the App without Internet Connection?

**Comment by Ein-Tim** [comment]
Update from my side @ndegendogo & @daimpi 

iPhone 6s | iPhone XR
------------ | -------------
![FA88FFDE-B4F6-490C-8F66-DE2A322D38C1](https://user-images.githubusercontent.com/67682506/97109915-621fdd80-16d6-11eb-81ec-6cb8435b2420.gif)|![RPReplay-Final1603625179](https://user-images.githubusercontent.com/67682506/97110497-efb0fc80-16d9-11eb-8384-8969fe93fabd.gif)
Works as expected | Doesn't work as expected

Please note that I still could not find any pattern when it works and when not, it just sometimes do and sometimes not 🤷‍♂️

**Comment by thomasaugsten** [comment]
For me it looks like the 6s has a cached app config and can start without internet the other phone has never received a app config and cannot start.

**Comment by Ein-Tim** [comment]
@thomasaugsten 
Seems unlikely to me because this XR is my phone which I use every day and I open CWA very often there... 
But it still could be because as far as I remember I never had this problem on the 6s so maybe its a prblem on the XR...
If it's ok I would watch the behavior of the iPhone 6s in the next days and if there everything works and other iPhone Users can not reproduce the behavior of my iPhone XR, this Issue can be closed 👍
Will report back in 3-5 days 👍
Thanks for your work @thomasaugsten

**Comment by ndegendogo** [comment]
@thomasaugsten today I also got this screen, it is  contradictory and really confusing (see screenshot).

 cwa 1.5.2(1) / iOS 14.1 / iPhone 8

How to reproduce (my guess):
- enter an area with no internet coverage
- start cwa **before** iOS is aware of this fact

After a few seconds the error message popped up (no internet access, see second screenshot).

**Comment by ndegendogo** [comment]
.

**Comment by ndegendogo** [comment]
Especially confusing is the fact that very different functions are all called "Risiko-Ermittlung":
- the logging of contacts / RPI => this function is still active
- the download of DEK keys, followed by the matching and risk assessment => this function of course needs internet

Suggestion: find a better wording for the different steps of "Risiko-Ermittlung", like:
Begegnungs-Aufzeichnung
Kontakt-Abgleich
Risiko-Berechnung

Also the recommendation is confusing:
"Aktivieren Sie die Risiko-Ermittlung" / "Risiko-Ermittlung einschalten" should be something like "Aktivieren Sie das Internet"

**Comment by dsarkar** [comment]
Dear @ndegendogo, @kbobrowski, @Ein-Tim,

I think this is has been resolved and we can close it?

Best wishes,
DS

---
Corona-Warn-App Open Source Team

**Comment by ndegendogo** [comment]
As @Ein-Tim pointed out [here](https://github.com/corona-warn-app/cwa-backlog/issues/1#issuecomment-687175144), the current message is a bit confusing: "Risiko-Ermittlung aktiv / Risiko-Ermittlung gestoppt" on the same user screen.

@Ein-Tim you have a very good overview over the tickets - is this confusing wording tracked on another ticket, or here?

**Comment by Ein-Tim** [comment]
@ndegendogo 

> you have a very good overview over the tickets - is this confusing wording tracked on another ticket, or here?

This Issue is tracked here AFAIK (since the confusing wording will be solved when this is solved). 

@dsarkar Could we please wait for the upcoming 1.7.0 release this week, I would like to do some additional testing if there are any situations the App still shows [this](https://user-images.githubusercontent.com/67682506/96738140-b49f8800-13be-11eb-868c-af02c08aea49.png) screen. If not this ticket can be closed then 👍

**Comment by ndegendogo** [comment]
But, actually, the root cause of the confusion is that they use the same wording "Risiko-Ermittlung" for two partial activities:  
1) capture and store RPI (= Erfassung / Aufzeichnung der Begegnungen)
2) matching and risk assessment (= Auswertung, Risiko-Berechnung),
and these have different hardware requirements (first needs Bluetooth, the second internet). 
So it was a good idea to split into two messages. Only the wording should be improved.

**Comment by Ein-Tim** [comment]
@ndegendogo 

Yes you are right, this is the root cause, but AFAIK (and my quick research over the Issues in the documentation Repo) your comment [here](https://github.com/corona-warn-app/cwa-backlog/issues/1#issuecomment-719793486) is the only one which mentions this problem. 

Also the explanation which the App gives under "Überblick" is even more confusing.
There the little box on the top "Risiko Ermittlung aktiv/gestoppt" is explained like this:
`[...] Ist sie aktiv, verden Begegnungen aufgezeichnet. [...]` 
That's true, but why is this called "Risiko Ermittlung" and not somethin like "Begegnungsaufzeichung"?

Then the big box (Risk Card) in the middle is introduced as: "Infektionsrisiko". Thats true, but only if the card is either green, red or gray. If the card is white the the "Infektionsrisiko" Box is suddenly showing somethin about the "Risiko-Ermittlung". Why? 

I think it would be the best if you (@ndegendogo) open an Issue in the Documentation Repo about this. (If you don't want to open one, I could, of course, also open one) 😀

Have a nice evening!

**Comment by ndegendogo** [comment]
> I think it would be the best if you open an Issue in the Documentation Repo about this. (If you don't want to open one, I could, of course, also open one) 😀

@Ein-Tim I think you are right. A new ticket, focussed on the confusing wording, seems a good idea.
If you don't mind - go ahead and write it 👍

**Comment by Ein-Tim** [comment]
Mhh @ndegendogo I've started to draft a new Issue but I after I thought about it again I think it is not nessecary. 
[This](https://user-images.githubusercontent.com/6118267/97754778-642fd500-1af8-11eb-8c38-5c275ef344d9.png) very confusing and contradictory screen was fixed (got heavily better) with 1.3.0 and above. 
AFAIK there are (at the moment) no other situations where the App shows this screen (or a similar contradictory screen like this).

See also the iOS Issue https://github.com/corona-warn-app/cwa-app-ios/issues/1497. 
There the App, in the little box, shows:
`Expsoure Logging Actvie`, which is correct, RPI are constantly recived/send/stored.
The big box shows:
`Exposure CHECK failed` (`Risiko-ÜBERPRÜFUNG fehlgeschlagen`), which is correct!
I hope you understand what I mean 😅

**Comment by ndegendogo** [comment]
@Ein-Tim yes, you are right. The messages in 1.6.1 that you linked look MUCH better.

**Comment by MikeMcC399** [comment]
@Ein-Tim 
I guess you are reporting about what you see on iOS, correct?

I can see EXPOSURE LOGGING ACTIVE on Android 1.6.1 when WiFi and mobile network are disabled (provided Bluetooth is enabled). So that is good.
I don't get any other message though. I think that Android would only complain if it couldn't download keys for a couple of days. It wouldn't give an error if the network is temporarily unavailable.

Anyway on Android I don't see any wrong status information being displayed.

So, are we saying that the issue is solved on both iOS and Android? I think, yes!

**Comment by MikeMcC399** [comment]
A related Android issue https://github.com/corona-warn-app/cwa-app-android/issues/1141 was confirmed closed two days ago.

**Comment by Ein-Tim** [comment]
@MikeMcC399 

> So, are we saying that the issue is solved on both iOS and Android? I think, yes!

Yes, but under 1.5.0 there still were some conditions in that the App showed this screen, with 1.6.0/1 we can't really test this because of https://github.com/corona-warn-app/cwa-app-ios/issues/1497. 
So it would be the best if this Issue does not yet get closed before we can test the behavior under 1.7.0 👍
